compiler/src/dmd/dtemplate.d

```diff
@@ -6261,27 +6261,40 @@ extern (C++) class TemplateInstance : ScopeDsymbol
             return false;
         }
 
+        // When no errors instantiating, this should only be called on
+        // the primary instantiation.
+        assert(this is inst || inst is null);
+
         if (global.params.allInst)
         {
             // Do codegen if there is an instantiation from a root module, to maximize link-ability.
-
-            // Do codegen if `this` is instantiated from a root module.
-            if (minst && minst.isRoot())
-                return true;
-
-            // Do codegen if the ancestor needs it.
-            if (tinst && tinst.needsCodegen())
+            static ThreeState needsCodegenAllInst(TemplateInstance tithis, TemplateInstance tinst)
             {
-                minst = tinst.minst; // cache result
-                assert(minst);
-                assert(minst.isRoot());
-                return true;
+                // Do codegen if `this` is instantiated from a root module.
+                if (tithis.minst && tithis.minst.isRoot())
+                    return ThreeState.yes;
+
+                // Do codegen if the ancestor needs it.
+                if (tinst && tinst.inst && tinst !is tinst.inst)
+                    tinst = tinst.inst;
```

Comment:
If `tinst` previously wasn't guaranteed to be the primary instance, then this might indeed fix existing issues wrt. missing symbols!

Reply:
Don't be so sure, `-allinst` is the only strategy that seems to consistently link programs on all the current regression reports I can see related to template culling.